File: react-native/components/chat/card-registry.ts
/**
 * Card Registry
 *
 * Maps session_message `kind` values to their rendering components.
 * When ChatTranscript encounters a non-text message, it looks up the
 * kind in this registry and renders the mapped component.
 *
 * Placeholder stubs are used for kinds whose underlying tools are not
 * yet wired up (weather, saved_route) — they render nothing until
 * replaced with real components.
 */

import type { ComponentType } from 'react'
import type { Id } from '../../../server/convex/_generated/dataModel'
import { LocationSearchCard } from './cards/location-search-card'
import { PlanningCard } from './cards/planning-card'
import { ReasoningCard } from './cards/reasoning-card'
import { ThinkingCard } from './cards/thinking-card'
import { RoutingCard } from './routing-card'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type CardAttachment =
  | { type: 'route_options'; routePlanId: Id<'route_plans'> }
  | {
      type: 'location_search'
      searchQuery: string
      results: {
        id: string
        name: string
        address: string
        types?: string[]
        location: { lat: number; lng: number }
        detourMinutes?: number
        distanceMeters?: number
      }[]
    }

export type CardProps = {
  message: {
    _id: Id<'session_messages'>
    createdAt: number
    status?: 'streaming' | 'running' | 'complete' | 'failed'
    content: string
    thinkingSteps?: {
      type: 'thinking' | 'tool_start' | 'tool_finish'
      toolName?: string
      summary: string
      detail?: string
      timestamp: number
    }[]
  }
  attachments: CardAttachment[]
  /** Called when the user taps a completed route card to view it on the map. */
  onViewOnMap?: () => void
}

export type CardKind =
  | 'routing_card'
  | 'weather_card'
  | 'saved_route_card'
  | 'reasoning'
  | 'thinking_card'
  | 'planning'
  | 'location_search_card'

// ---------------------------------------------------------------------------
// Placeholder stub — renders nothing until a real tool-backed card exists
// ---------------------------------------------------------------------------

const PlaceholderCard: ComponentType<CardProps> = () => null
PlaceholderCard.displayName = 'PlaceholderCard'

// ---------------------------------------------------------------------------
// Registry
// ---------------------------------------------------------------------------

export const CARD_REGISTRY: Record<CardKind, ComponentType<CardProps>> = {
  routing_card: RoutingCard as ComponentType<CardProps>,
  weather_card: PlaceholderCard, // TODO: replace when fetchWeather tool is real
  saved_route_card: PlaceholderCard, // TODO: replace when saveRoute tool is real
  reasoning: ReasoningCard,
  thinking_card: ThinkingCard,
  planning: PlanningCard,
  location_search_card: LocationSearchCard,
}

File: android/app/src/main/java/com/laneshadow/ui/components/chat/CardRegistry.kt
package com.laneshadow.ui.components.chat

import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.laneshadow.ui.components.chat.cards.ReasoningCard
import com.laneshadow.ui.components.molecules.ChatMessage
import com.laneshadow.ui.components.molecules.ChatMessageKind

/**
 * Card Registry
 *
 * Maps ChatMessageKind values to their rendering components.
 * When ChatTranscript encounters a non-text message, it looks up the
 * kind in this registry and renders the mapped component.
 *
 * Placeholder stubs are used for kinds whose underlying tools are not
 * yet wired up (weather, saved_route) — they render nothing until
 * replaced with real components.
 *
 * @param kind The type of card to render
 * @param message The chat message containing card data
 * @param modifier Modifier for the card component
 * @return The composable card content, or null if not implemented
 */
@Composable
fun renderCard(
    kind: ChatMessageKind?,
    message: ChatMessage,
    modifier: Modifier = Modifier,
): Unit? {
    // If kind is null or TEXT, return null — handled by caller (text rendering)
    if (kind == null || kind == ChatMessageKind.TEXT) {
        return null
    }

    // Dispatch to appropriate card renderer based on kind
    // All cards currently return null (not implemented) until their
    // underlying tools are wired up
    return when (kind) {
        ChatMessageKind.ROUTING_CARD -> null // TODO: Implement RoutingCard
        ChatMessageKind.WEATHER_CARD -> null // TODO: Implement WeatherCard
        ChatMessageKind.SAVED_ROUTE_CARD -> null // TODO: Implement SavedRouteCard
        ChatMessageKind.REASONING -> {
            ReasoningCard(message = message, modifier = modifier)
            Unit // Return Unit to satisfy nullable return type
        }
        ChatMessageKind.THINKING_CARD -> null // TODO: Implement ThinkingCard
        ChatMessageKind.PLANNING -> null // TODO: Implement PlanningCard
        ChatMessageKind.LOCATION_SEARCH_CARD -> null // TODO: Implement LocationSearchCard
        ChatMessageKind.TEXT -> null // Handled by caller
    }
}
